custom-plugin-decorators/remove-unused-tags/tags.js

```diff
@@ -0,0 +1,68 @@
+module.exports = {
+  id: "tags",
+  decorators: {
+    oas3: {
+      "no-unused-tags": ({ ignore }) => {
+        console.log("Checking for unused tags...");
+        return {
+          Root: {
```

Comment:
I believe it'd be much simpler to use `Operation` visitor instead of `Root` as this is the only possible place where the tags can be used. Something like this:

```js
const usedTags: new Set()

Operation: {
  enter(operation, ctx) {
    for (const tag of operation.tags) {
      usedTags.add(tag.toLowerCase())
    }
  }

```

And then you can iterate over `TagList` and remove anything that is not either in the `usedTags` set or in `tagsToIgnore` array. (Maybe it'd be even simpler to initiate the `unusedTags` set with the ignored tags at the beginning).

Reply:
I used `Root` because I need to read the top-level `tags` entry that defines the tags, then all the `paths` to see which tags are used, then operate on that top-level `tags` if there are any that aren't used.

Reply:
Yes. However, it'd be much easier to split it into two steps: first, to grab all the tags used in Operations on `enter`, and second, to remove all unused tags from the TagList on `leave`. Let me show you the idea (not tested though):

```js
({ ignore }) => {
  const usedTags = new Set(ignore?.map((tag) => tag.toLowerCase()));
  return {
    Root: {
      leave(root) {
        root.tags = root.tags.filter((tag) => usedTags.has(tag.name.toLowerCase()));
      },
      Operation: {
        enter(operation) {
          for (const tag of operation.tags) {
            usedTags.add(tag.toLowerCase());
          }
        },
      },
    },
  };
};
```

Reply:
Does this approach work because Root leave() will always be the last thing? I originally had separate operation and taglist parts, but I wasn't confident the visitor pattern would always order as I had expected. I will have another go at this with your suggestion, thanks!

Reply:
> Does this approach work because Root leave() will always be the last thing?

Yes, that's the idea.